File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    
    # 많이 플레이된 장르순 -> 장르 내에서 가장 많이 플레이된 노래 순으로 2개 수록
        # 같은 경우 idx가 작은 것부터 수록
    
    # 장르별 합산
    genre_dict = defaultdict(int)
    # 장르별 노래
    genre_song_dict = defaultdict(lambda : [])
    
    for i in range(len(plays)):
        genre = genres[i]
        song = i
        play = plays[i]
        
        genre_dict[genre] += play
        genre_song_dict[genre].append([song, play])
        
    # 장르별 순서 정렬
    for genre, song_list in genre_song_dict.items():
        genre_song_dict[genre] = [song_idx for song_idx, play in sorted(song_list, key = lambda x: -x[1])[:2]]
        
    print(genre_song_dict)

    for genre, play in sorted(list(genre_dict.items()), key = lambda x: -x[1]):
        answer += genre_song_dict[genre]
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
// 속한 노래들의 총 재생 횟수가 많은 순으로 수록 2개
// 장르 내에서는 재생 횟수가 많은 노래 순으로 수록 2개
    // 재생 횟수가 같다면 고유 번호가 낮은 노래를 먼저 수록
    // 고유번호는 idx

function genreSort(a, b){
    const [aPlay, aIndex] = a;
    const [bPlay, bIndex] = b;
    
    if(aPlay != bPlay) return bPlay - aPlay;
    else return aIndex - bIndex;
}

function solution(genres, plays) {
    let answer = [];
    
    // 장르별로 [노래, 고유번호]를 취합
    const genreMap = new Map();
    
    for(let i = 0; i < plays.length; i++){
        genre = genres[i];
        play = plays[i];
        
        if(genreMap.has(genre)){
            genreMap.get(genre).push([play, i]);
        }else{
            genreMap.set(genre, [[play, i]]);
        }
    }
    
    // 장르별 배열에 정렬
    
    genreMap.forEach((item) => item.sort(genreSort));
    
    // 장르별 배열에 노래 횟수 취합
    
    genreCountMap = new Map();
    
    for(const [genre, songArr] of genreMap){
        genreCountMap.set(genre, songArr.reduce((acc, [play, idx]) => acc + play, 0));
    }
    
    return Array.from(genreCountMap.entries()).sort((a, b) => b[1] - a[1]).map(item => item[0]).map(genre => genreMap.get(genre).slice(0, 2).map(([play, idx]) => idx)).flat()
}